hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/HoodieTimelineArchiver.java

```diff
@@ -475,7 +485,24 @@ private Stream<HoodieInstant> getCommitInstantsToArchive() throws IOException {
                       HoodieTimeline.compareTimestamps(s.getTimestamp(), LESSER_THAN, instantToRetain.getTimestamp()))
                   .orElse(true)
           );
-      return instantToArchiveStream.limit(commitTimeline.countInstants() - minInstantsToKeep);
+      List<HoodieInstant> instantsToArchive = instantToArchiveStream.limit(commitTimeline.countInstants() - minInstantsToKeep).collect(Collectors.toList());
+      // If cleaner is based on hours, lets ensure hudi does not archive commits yet to cleaned by the cleaner.
+      if (config.getCleanerPolicy() == HoodieCleaningPolicy.KEEP_LATEST_BY_HOURS && !instantsToArchive.isEmpty()) {
```

Comment:
this is another special rule added to the codebase and increased logic complexity. how can we design it better so that cleaner policy can fit nicely with archiver. for e.g., cleaner policy can have an API say `canArchive()` and archiver just use it. we need to design better APIs rather than adding conditions to just make things work.

Reply:
yeah. I agree. but the guard conditions differ slightly from one policy to another. for eg, incase of num_commits, we can validate just w/ right configs. but if its based on hours, since archival is based on num of commits, we had to defer this check to when we actually do the archival. I do have a follow up patch where we will do some policy agnostic check in archival. but for this, do you think we need to think about adding a new api in cleaner code.

Reply:
do you have a follow up work/ticket for solving cleaner/archiver entanglement? let's link the jiras?